Add FXIOS-9853 connection between redux and navigation

petruSt24 opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9853)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/21632)

## :bulb: Description
<!--- Describe your changes so the reviewer can understand the context and decisions made for your work -->
Connected the redux actions with the navigation towards the other controllers
## :pencil: Checklist
You have to check all boxes before merging
- [X] Filled in the above information (tickets numbers and description of your work)
- [X] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [X] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

cyndichin commented:
@petruSt24 seems like there's some conflicts, let me know when it's ready for review!

petruSt24 commented:
@cyndichin I've updated the PR with more bugfixes, rebased it, organised the files better and I intend to move the navigation code into the nav coordinator and unify it with the iPad navigation on the next one, but at least this PR would unblock QA from testing the tracking protection flow on iPhone

cyndichin commented on firefox-ios/Client/Frontend/TrackingProtection/BlockedTrackersHelpers/BlockedTrackersTableView.swift:
```diff
@@ -48,7 +48,7 @@ class BlockedTrackersTableView: UITableView,
         snapshot.appendSections([0])
         snapshot.appendItems(items, toSection: 0)
-        diffableDataSource?.apply(snapshot, animatingDifferences: true)
+        diffableDataSource?.apply(snapshot, animatingDifferences: false)
```
curious - why false here? I'm also working with diffable datasource for the homepage rebuild project.

petruSt24 replied:
https://github.com/user-attachments/assets/2ab07b7e-7c57-4518-8b6d-b32770c25b0c

trying to lower the number of epileptic seizures the users might experience because of how refreshing the number while the URL loads is animating the differences

petruSt24 replied:
https://github.com/user-attachments/assets/16932b72-6367-4032-983a-3876d86d11d9

maybe an animation could look nice though, but certainly not the one the SDK provides

cyndichin replied:
ah thank you for explaining! yes the first video is not great at all 🤣

cyndichin commented on firefox-ios/Client/Frontend/TrackingProtection/TrackigProtectionRedux/TrackingProtectionAction.swift:
```diff
@@ -17,6 +17,8 @@ enum TrackingProtectionActionType: ActionType {
     case tappedShowTrackingProtectionDetails
     case tappedShowBlockedTrackers
     case tappedShowClearCookiesAlert
+    case resetReduxStore
```
can we rename this as a user / api action as opposed to the consequence of what occurs?

cyndichin commented on firefox-ios/Client/Frontend/TrackingProtection/TrackigProtectionRedux/TrackingProtectionState.swift:
```diff
@@ -134,11 +201,13 @@ struct TrackingProtectionState: StateType, Equatable, ScreenState {
+                shouldUpdateBlockedTrackerStats: false
             )
         default:
             return state
```
we should be returning a new instance for our default, we discussed something during our team meeting, probably should have it documented somewhere. `return TrackingProtectionState(...)`

cyndichin commented on firefox-ios/Client/Frontend/TrackingProtection/TrackingProtectionDetailsViewController.swift:
```diff
@@ -18,14 +18,24 @@ struct TrackingProtectionDetailsModel {
         static let baseCellHeight: CGFloat = 44
         static let baseDistance: CGFloat = 20
-        static let bottomDistance: CGFloat = 350
+        static let bottomDistance: CGFloat = 0
```
do we still need this value, if it's 0?

cyndichin commented:
> @cyndichin I've updated the PR with more bugfixes, rebased it, organised the files better and I intend to move the navigation code into the nav coordinator and unify it with the iPad navigation on the next one, but at least this PR would unblock QA from testing the tracking protection flow on iPhone

thanks for the clarity! looks a lot better! thank you for organizing and the details. left a couple of initial comments. Let me know what you think!

petruSt24 commented on firefox-ios/Client/Frontend/TrackingProtection/BlockedTrackersTableController.swift:
```diff
@@ -183,6 +187,16 @@ class BlockedTrackersTableViewController: UIViewController,
 
+    // MARK: Header Actions
+    private func setupHeaderViewActions() {
+        navigationView.backToMainMenuCallback = { [weak self] in
+            self?.navigationController?.popToRootViewController(animated: true)
+        }
+        navigationView.dismissMenuCallback = { [weak self] in
+            self?.navigationController?.dismissVC()
+        }
```
that navigationView is a common component, I agree that dispatching actions should be used in the callback, but removing the callback altogether would affect the other places it's used as well (for now the menu redesign)

cyndichin commented on firefox-ios/Client.xcodeproj/project.pbxproj:
```diff
@@ -16106,7 +16138,7 @@
 				0B3F8C5E2CA4471C00DB5367 /* EditBookmarkViewModel.swift in Sources */,
 				8AD0110D2CB42FE700368BF3 /* HeaderState.swift in Sources */,
-				D0B9483422A03468002F4AA1 /* BookmarkDetailPanel.swift in Sources */,
+				D0B9483422A03468002F4AA1 /* LegacyBookmarkDetailPanel.swift in Sources */,
```
this seems unrelated to ETP, do we want to revert this change?

cyndichin commented on firefox-ios/Client/Frontend/TrackingProtection/TrackigProtectionRedux/TrackingProtectionState.swift:
```diff
@@ -134,11 +201,24 @@ struct TrackingProtectionState: StateType, Equatable, ScreenState {
+                showTrackingProtectionSettings: false,
+                trackingProtectionEnabled: state.trackingProtectionEnabled,
+                connectionSecure: state.connectionSecure,
+                showDetails: false,
+                showBlockedTrackers: false,
+                showsClearCookiesAlert: false,
```
nit: instead of all the booleans, can follow similar to how we show certain views using `DisplayType` in `BrowserViewControllerState`?

petruSt24 replied:
yes it would be much cleaner that way, but I wanted to do the final navigation cleanup on a different PR to avoid making this one too big (it's already quite big) and to better organise the coordinator on the iPhone/iPad differences

cyndichin replied:
sounds good, thank you!

cyndichin reviewed (commented):
thanks @petruSt24 looks good now! one minor comment on the renaming of bookmarks

cyndichin reviewed (approved):
thanks for making the updates + addressing comments!
